File: adcs-frontend/src/api/department.ts
import axiosClient from './axiosClient';

// Định nghĩa kiểu dữ liệu payload gửi lên backend
export interface DepartmentPayload {
  id?: number | null;
  name: string;
  code?: string | null;
  description?: string | null;
  ten_viet_tat?: string | null;
  ten_hien_thi?: string | null;
  loai_don_vi?: string | null;
  cap_don_vi?: string | null;
  level_number?: number | null;
  is_formal?: boolean | null;
  has_seal?: boolean | null;
  parent_name?: string | null;
  child_count?: number | null;
  parent_id?: number | null;
}

// Định nghĩa kiểu dữ liệu trả về từ backend
export interface DepartmentResponse extends DepartmentPayload {
  department_id: string;
  created_at: string;
}

export interface DepartmentTreeResponse extends DepartmentResponse {
  children: DepartmentTreeResponse[];
}

export const departmentApi = {
  getAll: (): Promise<DepartmentTreeResponse[]> => 
    axiosClient.get('/departments'),

  create: (data: DepartmentPayload): Promise<DepartmentResponse> => 
    axiosClient.post('/departments', data),

  update: (id: string, data: DepartmentPayload): Promise<DepartmentResponse> => 
    axiosClient.put(`/departments/${id}`, data),

  delete: (id: string): Promise<{ message: string }> => 
    axiosClient.delete(`/departments/${id}`),
};

File: adcs-backend/app/schemas/department.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from uuid import UUID
from datetime import datetime

class DepartmentBase(BaseModel):
    name: str
    id: Optional[int] = None
    code: Optional[str] = None
    description: Optional[str] = None
    ten_viet_tat: Optional[str] = None
    ten_hien_thi: Optional[str] = None
    loai_don_vi: Optional[str] = None
    cap_don_vi: Optional[str] = None
    level_number: Optional[int] = None
    is_formal: Optional[bool] = None
    has_seal: Optional[bool] = None
    parent_name: Optional[str] = None
    child_count: Optional[int] = 0
    parent_id: Optional[int] = None

class DepartmentCreateRequest(DepartmentBase):
    pass

class DepartmentUpdateRequest(BaseModel):
    name: Optional[str] = None
    id: Optional[int] = None
    code: Optional[str] = None
    description: Optional[str] = None
    ten_viet_tat: Optional[str] = None
    ten_hien_thi: Optional[str] = None
    loai_don_vi: Optional[str] = None
    cap_don_vi: Optional[str] = None
    level_number: Optional[int] = None
    is_formal: Optional[bool] = None
    has_seal: Optional[bool] = None
    parent_name: Optional[str] = None
    child_count: Optional[int] = None
    parent_id: Optional[int] = None

class DepartmentResponse(DepartmentBase):
    department_id: UUID
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class DepartmentTreeResponse(DepartmentResponse):
    children: List["DepartmentTreeResponse"] = []

DepartmentTreeResponse.model_rebuild()

class MessageResponse(BaseModel):
    message: str

class RegionUnitResponse(BaseModel):
    ma_don_vi: Optional[str] = None
    id: Optional[int] = None
    ten_hien_thi: Optional[str] = None
    nhom: Optional[str] = None
    mo_ta: Optional[str] = None

class RegionSearchResponse(BaseModel):
    region: str
    label: str
    units: List[RegionUnitResponse] = []